indexed_gzip/zran.c

```diff
@@ -2384,3 +2384,450 @@ eof:         return ZRAN_READ_EOF;
+    f_ret = fwrite(&index->npoints, sizeof(index->npoints), 1, fd);
+
+    if (ferror(fd)) goto fail;
+    if (f_ret != 1) goto fail;
+
+    /*
+     * Initialize point to the first element of the list, and list_end to the
+     * end of the list.
+     */
+    point    = index->list;
+    list_end = index->list + index->npoints;
+
+    /* Write all points iteratively. */
+    while (point < list_end) {
+
+        /*
+         * Below, it's preferred to write members of points elementwise to
+         * avoid handling struct alignment and padding issues.
+         */
+
+        /* Write compressed offset, and check for errors. */
+        f_ret = fwrite(&point->cmp_offset, sizeof(point->cmp_offset), 1, fd);
+
+        if (ferror(fd)) goto fail;
+        if (f_ret != 1) goto fail;
+
+        /* Write uncompressed offset, and check for errors. */
+        f_ret = fwrite(&point->uncmp_offset, sizeof(point->uncmp_offset), 1, fd);
+
+        if (ferror(fd)) goto fail;
+        if (f_ret != 1) goto fail;
+
+        /* Write bit offset, and check for errors. */
+        f_ret = fwrite(&point->bits, sizeof(point->bits), 1, fd);
+
+        if (ferror(fd)) goto fail;
+        if (f_ret != 1) goto fail;
+
+        /*
+         * Data is NULL for the first element of the list, and therefore it is
+         * exported for all elements except the first one.
+         */
+        if(point != index->list) {
+
+            /* Write checkpoint data, and check for errors. */
+            f_ret = fwrite(point->data, index->window_size, 1, fd);
+
+            if (ferror(fd)) goto fail;
+            if (f_ret != 1) goto fail;
+        }
+
+        #ifdef ZRAN_VERBOSE
+        zran_log("zran_export_index: (cmp: %lu, uncmp: %lu, bits: %u, ",
+                 point->cmp_offset,
+                 point->uncmp_offset,
+                 point->bits);
+
+        /* First point supposed to not have a checkpoint window. */
+        if(point->data != NULL) {
+
+            /* Print first and last three bytes of the checkpoint window. */
+            zran_log("window: [%02x %02x %02x...%02x %02x %02x])\n",
+                     point->data[0],
+                     point->data[1],
+                     point->data[2],
+                     point->data[index->window_size - 3],
+                     point->data[index->window_size - 2],
+                     point->data[index->window_size - 1]);
+        } else {
+            zran_log("window: NULL)\n");
+        }
+        #endif
+
+        /* Done with this point. Proceed to next one. */
+        point++;
+    }
+
+    /*
+     * It is important to flush written file when done, since underlying file
+     * descriptor can be closed by Python code before having a chance to flush.
+     */
+    fflush(fd);
+
+    return ZRAN_EXPORT_OK;
+
+fail:
+    return ZRAN_EXPORT_WRITE_ERROR;
+
+}
+
+/*
+ * Load checkpoint information from file fd to index. File should be opened in
+ * binary read mode.
```

Comment:
You could use the `is_readonly` function (defined at the top of the file) to check correct permissions.